src/components/ProductForm/ProductFormContext.tsx

```diff
@@ -0,0 +1,146 @@
+const priceIsValid = (priceString: string): boolean => {
+  const price = parseFloat(priceString);
+  return !isNaN(price) && Number(price) >= PRICE_MIN && price <= PRICE_MAX;
+};
+
+const DESCRIPTION_MAX_LENGTH = 500;
+const VALIDATION_MSG_DESC = 'Description is required and must be less than 500 characters in length.';
+
+const descriptionIsValid = (description: string): boolean => {
+  if (!description) {
+    return false;
+  }
+  return description.length <= DESCRIPTION_MAX_LENGTH;
+};
+
+interface IFormProduct {
+  name: string;
+  desc: string;
+  price: number;
+  stock: number;
+  tags: number[];
+}
+
+interface IProductFormContext {
+  startEditing: (product: ProductDto) => void;
+  startNew: () => void;
+  useFormProduct: [IFormProduct, (newProduct: IFormProduct) => void];
+  closeForm: () => void;
+  saveForm: () => void;
+}
+
+const ProductFormContext = createContext<IProductFormContext>({
+  startEditing: () => null,
+  startNew: () => null,
+  useFormProduct: [{ name: '', desc: '', price: 0, stock: 0, tags: [] }, () => null],
+  closeForm: () => null,
+  saveForm: () => null,
+});
+
+interface Props {
+  children: React.ReactNode;
+}
+
+export const ProductFormContextProvider: React.FC<Props> = ({ children }) => {
+  const [initialProduct, setInitialProduct] = useState<ProductDto | undefined>(undefined);
+  const [formOpen, setFormOpen] = useState<boolean>(false);
+  const { createProduct } = useProducts();
+  const [formProduct, setFormProduct] = useState<IFormProduct>({ name: '', desc: '', price: 0, stock: 0, tags: [] });
+  const [validationErrors, setValidationErrors] = useState<string[]>([]);
+  const toastId = useRef(0);
+
+  const editProduct = (newProduct: ProductDto): void => {
+    setFormOpen(true);
+    setInitialProduct(newProduct);
+    setFormProduct({ name: newProduct.name, desc: newProduct.description, price: newProduct.price, stock: 1, tags: [] });
+  };
+
+  const newProduct = (): void => {
+    setFormOpen(true);
+    setInitialProduct(undefined);
+    setFormProduct({ name: '', desc: '', price: 0, stock: 0, tags: [] });
+  };
+
+  const closeForm = (): void => {
+    setInitialProduct(undefined);
+    setFormOpen(false);
+  };
+
+  const handleErrorClear = (): void => {
+    setValidationErrors([]);
+    createProduct.reset();
+    toast.dismiss(toastId.current);
+  };
+
+  const saveForm = (): void => {
+    handleErrorClear();
+    if (!nameIsValid(formProduct.name)) {
+      setValidationErrors((errors) => [...errors, VALIDATION_MSG_NAME]);
+    }
+    if (!descriptionIsValid(formProduct.desc)) {
+      setValidationErrors((errors) => [...errors, VALIDATION_MSG_DESC]);
+    }
+    if (!priceIsValid(formProduct.price.toString())) {
+      setValidationErrors((errors) => [...errors, VALIDATION_MSG_PRICE]);
+    }
+
+    if (nameIsValid(formProduct.name) && descriptionIsValid(formProduct.desc) && priceIsValid(formProduct.price.toString())) {
```

Comment:
I don't like doing a second round of checks when you've already called these functions, I would just create a `var error = false` at the top then set it to true if any of the validation fails and check that here

Reply:
You could probably actually go further and make a generic function that passes in the validation function, then does the check, calls setValidationErrors if failed, and returns a boolean result. Then you'd only need one line:

`if (valid(() => nameIsValid(formProduct.name), VALIDATION_MSG_NAME)
&& valid(() => descriptionIsValid(formProduct.desc), VALIDATION_MSG_DESC)
&& valid(() => priceIsValid(formProduct.price), VALIDATION_MSG_PRICE)) {
...  
}`

Reply:
This is Gramham's old code that I've minimally modified to keep working and is outside the scope of the ticket, validation probably needs to be a custom hook or something. A problem for BENCH-290 when all the functionality is added.